### PR TITLE
Am error handling testing

### LINKED ISSUE #4: Iteration 1: Error Handling
Add error handling functionality for any bad requests that the server may give

### LINKED ISSUE #5: Iteration 2: Error Handling
Add error handling functionality for any bad requests that the server may give

### LINKED ISSUE #6: Iteration 3: Error Handling
Add error handling functionality for any bad requests that the server may give

### LINKED ISSUE #7: Iteration 4: Error Handling
Add error handling functionality for any bad requests that the server may give

### LINKED ISSUE #8: Iteration 5: Error Handling
Add error handling functionality for any bad requests that the server may give

### LINKED ISSUE #19: Iteration 3: Movie "Show Page" Testing
test all functionality possible, sad path, etc.

### LINKED ISSUE #26: Testing: Async Functionality
Asynchronous tests should cover cover ALL functionality including:

- [ ]  happy paths

- [ ] multiple sad paths

### LINKED ISSUE #27: Iteration 1: Testing
Create tests for functionality of User Info, Fetching movies from API & movie display

### LINKED ISSUE #44: Iteration 4: Testing
Test everything!

- [ ] Display

- [ ] Functionality

- [ ] Sad path

- [ ] Edge cases

### LINKED ISSUE #45: Iteration 4: Error handling

### LINKED ISSUE #46: Iteration 5: Error Handling

### LINKED ISSUE #47: Iteration 5: Change Rating
The only way to change a rating is to DELETE the old rating & submit a new rating

Functionality must be that the user is editing their rating by deleting an old rating number & submitting a new rating number

### LINKED ISSUE #50: Iteration 5: Change rating
The only way to change a rating: 

- [ ] the original rating must be deleted

- [ ] new rating must be submitted

### LINKED ISSUE #51: Add proptypes

### LINKED ISSUE #52: Stars should not display on movie view when user is not signed in
- [ ] Stars should ONLY display on movie view when a user is signed in

- [ ] MovieView: replace "add a rating" with "sign in to add your rating"

### LINKED ISSUE #53: MovieView: Stars should only show when user signed in
Stars should only show on movie view when a user is signed in

### LINKED ISSUE #55: Refactor: manipulating history on MovieCard
Need to refactor per rubric. Maybe change to link?

### LINKED ISSUE #59: Fetch Bug in getUserRatings
Need to add a condition to not invoke the fetch if there is no user logged in.

### PR DESCRIPTION
**Pull request by aemcdonald**
## Description
Minor styling changes
Add error handling in App, all api methods
Added sad path test for login
Added test for user rating, user prompt in MovieCard
Removed history from MovieCard & updated test to reflect changes
Modify affected tests to pass due to bug fixes (all tests currently passing)
Added error handling in MovieGrid, added test to cover
Added PropTypes to all components & installed dependency
Adjusted minor linter errors
Added test for Rating component
Refactored Rating from class component to functional component
Added conditional rendering for Rating in MovieView
Added changeRating method in API calls
Added test to display user rating in MovieView
Minor changes to spacing, added semicolons
Created changeRating in apiCalls to delete user rating & added response.ok conditional to address json parsing bug
Fix bug with logout button in App.js
Added test for logout button functionality in App
Change key to index in MovieCard to fix bug

## Motivation and Context
Closes #27 
Closes #8 
Closes #46 
Closes #45 
Closes #7 
Closes #44 
Closes #59 
Closes #4 
Closes #6 
Closes #53 
Closes #52 
Closes #47 
Closes #50 
Closes #26 
Closes #51 
Closes #19 
Closes #5 
Closes #55 

## How Has This Been Tested?
All changes adequately covered with robust testing using jest
All tests currently passing

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/ApiCalls.js [hunk 1]
```diff
@@ -43,8 +42,18 @@ const ApiCalls = {
       body: JSON.stringify(ratingInfo)
     })
     .then(response => response.json())
-    .then(data => console.log('success:', data))
+    .then(data => data)
     .catch(err => console.log('failed', err.message))
+  },
+
+  changeRating(userId, ratingId) {
```
Comment (Kristiannmiller) [inline src/ApiCalls.js:49]: Good fix here. This takes care of that .json error we kept seeing

File: src/App/App.js [hunk 2]
```diff
@@ -13,25 +13,32 @@ export class App extends Component {
...
     this.setState({ user }, () => {this.props.history.push('/')})
   }
 
-  handleLoginSubmit = (userInfo) => {
-    ApiCalls.postUserLogin(userInfo)
-    .then(data => this.updateCurrentUser(data))
+  handleLoginSubmit = async (userInfo) => {
+    const loginResult = await ApiCalls.postUserLogin(userInfo)
+    if (loginResult.error) {
+      this.setState({error: loginResult.error})
```
Comment (Kristiannmiller) [inline src/App/App.js:28]: Nice error handling here. I think showing errors in state might be a good way to go for displaying the error to the user.

File: src/App/App.js [hunk 3]
```diff
@@ -53,11 +60,11 @@ export class App extends Component {
         <Route
           path='/login'
           render={(props) => (
-            <Login {...props} handleSubmit={this.handleLoginSubmit} />
+            <Login {...props} handleSubmit={this.handleLoginSubmit} errorMessage={this.state.error} />
```
Comment (Kristiannmiller) [inline src/App/App.js:63]: We've got some nice error handling going on in this PR!

File: src/App/App.test.js [hunk 4]
```diff
@@ -1,9 +1,40 @@
...
+import { render, screen, waitFor } from '@testing-library/react';
+import { BrowserRouter, Router } from 'react-router-dom';
+import { createMemoryHistory } from 'history';
 import { App } from './App';
+import ApiCalls from '../ApiCalls.js';
+import userEvent from '@testing-library/user-event';
+jest.mock('../ApiCalls.js');
+
 
 describe('App', () => {
```
Comment (Kristiannmiller) [inline src/App/App.test.js:11]: I'm a little unclear on this, but should all of our integration tests belong in the App test? It feels a bit sparse to me, but I could be wrong. Might want to get a second opinion on this.

File: src/Login/Login.js [hunk 5]
```diff
@@ -66,3 +64,8 @@ class Login extends Component {
 }
 
 export default Login;
+
+Login.propTypes = {
```
Comment (Kristiannmiller) [inline src/Login/Login.js:68]: almost forgot about PropTypes! Good catch!

File: src/Login/Login.test.js [hunk 6]
```diff
@@ -30,22 +27,39 @@ describe('Login', () => {
...
+    userEvent.click(screen.getByText('Login'))
     userEvent.type(screen.getByPlaceholderText('Email'), 'olivia@turing.io')
     userEvent.type(screen.getByPlaceholderText('Password'), 'pword')
     userEvent.click(screen.getByText('Login'))
 
     const welcomeMessage = await waitFor(() => screen.getByText('Welcome, Olivia!'))
     expect(welcomeMessage).toBeInTheDocument();
   })
+
+  it('should display an error message if user name or password is incorrect', async () => {
```
Comment (Kristiannmiller) [inline src/Login/Login.test.js:50]: Nice sad path here! The error handling on Login is really user-friendly.

File: src/MovieCard/MovieCard.js [hunk 7]
```diff
@@ -1,20 +1,26 @@
...
-    <section className='movieCard' onClick={() => {
-      history.push(`/${movie.id}`)}}>
+    <section className='movieCard'>
       <h1 className='cardTitle'>{movie.title}</h1>
       <h4>Average Rating: {parseFloat(movie.average_rating).toFixed(1)}</h4>
       <h4>{movie.userRating && `Your Rating: ${movie.userRating.rating}`}</h4>
       <h4>{!movie.userRating && userId && `Click to add rating`}</h4>
       <h4>{!movie.userRating && !userId && `Sign in to add your rating`}</h4>
-      <img className='poster' src={movie.poster_path} alt={movie.title + ' movie poster'}/>
+      <Link to={`/${movie.id}`}>
```
Comment (Kristiannmiller) [inline src/MovieCard/MovieCard.js:14]: I like the link on the image instead of on the entire card. It's a bit more intuitive. Can we possibly implement this into the logo at the top of page as well to get it to route back to the homepage? just a thought!

File: src/MovieGrid/MovieGrid.test.js [hunk 8]
```diff
@@ -26,4 +31,19 @@ describe('MovieGrid', () => {
     expect(movie2).toBeInTheDocument();
     expect(movie3).toBeInTheDocument();
   });
+
+  it('should render an error if no movies are found', async () => {
```
Comment (Kristiannmiller) [inline src/MovieGrid/MovieGrid.test.js:35]: Another great error handling and sad path test

File: src/Rating/Rating.css [hunk 9]
```diff
@@ -8,3 +8,7 @@
   display: flex;
   justify-content: center;
 }
+
+:focus {
```
Comment (Kristiannmiller) [inline src/Rating/Rating.css:12]: Glad we were able to fix this css issue with the box showing up around our stars. The stars are a little tricky. Maybe there is a better option for future implementation.

File: src/Rating/Rating.test.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+import ApiCalls from '../ApiCalls.js';
+import Rating from './Rating.js';
+import MovieView from '../MovieView/MovieView.js';
+import { render, screen, waitFor } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import '@testing-library/jest-dom';
+import { BrowserRouter } from 'react-router-dom';
+jest.mock('../ApiCalls');
+
+describe('Rating', () => {
```
Comment (Kristiannmiller) [inline src/Rating/Rating.test.js:12]: This is a lot of testing for a small functional component. Did we go a little overboard, or should some of these tests maybe live on another component?

**Comment by Kristiannmiller** [comment]
This PR looks great! I think we covered all of our error handling and testing here. I am wondering if maybe some of our tests need to be moved onto different components than the ones they're in. It seems like Rating is a little long while App is a little short and while it feels like it should be the opposite, the tests seem in the right place to me. Maybe we should get a second opinion on the testing locations.

File: src/App/App.test.js (hunk 4 above)
Comment (aemcdonald) [reply src/App/App.test.js:11]: I completely agree. I am getting the feeling that maybe all integrations tests belong in App? We should definitely seek some clarification on this from our instructors & refactor if we have time before the due date.

File: src/MovieCard/MovieCard.js (hunk 7 above)
Comment (aemcdonald) [reply src/MovieCard/MovieCard.js:14]: Oh that is an excellent idea! We should definitely implement clicking the logo to take the user back to the home page.

File: src/Rating/Rating.test.js (hunk 10 above)
Comment (aemcdonald) [reply src/Rating/Rating.test.js:12]: That's an excellent point, I think that the extra tests are necessary to cover all the user functionality but maybe some are considered integration tests & therefore they should live elsewhere?
